doc/form.1
.TH FORM 1 "2018-10-04"
.SH NAME
FORM \- Symbolic manipulation system
.SH SYNOPSIS
.B form
.RB [
.IR options 
]
.IR inputfile
.SH DESCRIPTION
.PP
FORM is a symbolic manipulation system. The \fBform\fR command reads a text file
(which should have a name that ends with the extension \fB.frm\fR, or \fB-\fR
to read the standard input) containing
definitions of mathematical expressions as well as statements that tell it how
to manipulate these expressions. It is widely used in the theoretical particle
physics community, but it is not restricted to applications in this specific
field.
.PP
\fBtform\fR is the threaded version using POSIX Threads.
.PP
\fBparform\fR is the multiprocessing version using MPI.
.SH OPTIONS
.TP
.BR "-"
Read the input from stdin instead of from a file.
.TP
.BR "-c"
Error checking only. Notice that this will not work properly if there are conditionals in the
preprocessor phase that depend on results obtained at earlier stages of the program.
.TP
.BR "-C"
Next argument is the filename of the log file. The default is the filename of the FORM script
with the .log extension.
.TP
.BR "-d, -D"
Next argument/option is the name of a preprocessor variable that will be defined before the
run starts. A specific value can be assigned with the syntax
\fB-d\fR\ \fIVARIABLENAME\fR=\fIVALUE\fR.
The
default value is 1.
.TP
.BR "-f"
Output goes only to log file.
.TP
.BR "-F"
Output only to log file. Further like \fB-L\fR or \fB-ll\fR.
.TP
.BR "-h"
Wait for some key to be touched before finishing the run. Basically only for some old window
based systems.
.TP
.BR "-install"
Install the package specified by the next argument.
.TP
.BR "-I"
Next argument/option is the path of a directory for include, procedure and subroutine files.
.TP
.BR "-l"
Make a regular log file.
.TP
.BR "-ll, -L"
Make a log file without intermediate statistics.
.TP
.BR "-M"
Put the PID (process identifier) in the name of the temporary files. This makes for longer
names, but gives a better guarantee of uniqueness. If a file with the created name exists
already it will be overwritten. This option is for when several instances of FORM are started
at nearly the same time as can happen from minos or make (with the make -j option).
.TP
.BR "-p"
Next argument/option is the path of a directory for input, include, procedure and subroutine
files.
.TP
.BR "-pipe"
Indicates that FORM is started up as the receiving end of a pipe. Action will be taken to
set up the proper communication channels.
.TP
.BR "-q, -si"
Quiet option. Only output expressions are printed.
.TP
.BR "-R"
Recover from a crash.
.TP
.BR "-s"
Next argument/option is the path of a directory for a setup file.
.TP
.BR "-S"
Next argument/option is the name of a setup file.
.TP
.BR "-t"
Next argument/option is the path of a directory for temporary files.
.TP
.BR "-ts"
Next argument/option is the path of a directory for temporary sort files.
.TP
.BR "-T"
Puts FORM in a mode in which the maximum totalsize is measured and printed at the end of
the program.
.TP
.BR "-v"
Only the version will be printed. The program terminates immediately after it.
.TP
.BR "-w"
This should be followed immediately by a number without any space. The number
indicates the number of worker threads for \fBtform\fR. All other versions of
FORM ignore this parameter.
.TP
.BR "-W"
Turn on the wall-clock time mode in the statistics.
.TP
.BR "-z"
The number following is a timelimit for the program in second.
.TP
.BR "-Z"
Removes the \fB.str\fR file on crash, whatever its contents. Under
ordinary circumstances at a crash a \fB.str\fR file will not be removed if
it has a nonzero content.
.TP
.BR "-y"
Run only the preprocessor and dump its output.
.SH ENVIRONMENT
.TP
\fBFORMPATH\fR
The directory in which FORM will look for procedures and header files, assuming it cannot
find them in the current directory.
.TP
\fBFORMTMP\fR
The directory in which FORM will make its temporary files.
.TP
\fBFORMTMPSORT\fR
The directory in which FORM will make its temporary sort files.
.TP
\fBFORMSETUP\fR
The full path and name of a setup file.
.SH SEE ALSO
.TP
\fBhttps://www.nikhef.nl/~form/\fR
The FORM home site.
.TP
\fBhttps://github.com/vermaseren/form/\fR
The repository on GitHub.


parform/ChangeLog
30 January 2004 Mikhail Tentyukov:

Here I summarize all changes I have made before this date.
All the code fragments I removed are encased in 
#ifdef REMOVEDBY_MT
...
#endif
All my changes are enclosed with marks 
[date mt]:
...
:[date mt]
e.g.
   /*[25nov2003 mt]:*/
   /* PF_outterms++;*/
   PF.goutterms++;
   /*:[25nov2003 mt]*/

The most important:

[17sep2003 mt]: Bugfix: on/off parallel takes effect only starting from the 
next module. The reason: flag AC.mparallelflag was set according to 
AC.parallelflag in the function IniModule. The solution: see comments 
[17sep2003 mt] in functions CoOff and CoOn, file compcomm.c

[05nov2003 mt]: Bugfix: after some expression was processed in parallel, 
there is no possibility to compile module with another expression. The reason:
after a flag PF.parallel is set to 1, functions  FlushOut and PutOut in slaves 
invoked at compile time send a mess to the master. The solution: PF.parallel=0
in IniModule, see pre.c

[06nov2003 mt]: To allow usage of RHS expressions I force ParFORM to switch to
sequential execution of moduli with RHS expressions. I introduced a counter 
AC.NumberOfRhsExprInModule which counts hom many times RHS expressions occur.
The counter is incremented in the function tokenize in the file token.c

[13nov2003 mt]: Bugfix: The function Processor worked incorrectly
with the following values of e->status:
UNHIDELEXPRESSION UNHIDEGEXPRESSION SKIPLEXPRESSION
SKIPGEXPRESSION HIDELEXPRESSION HIDEGEXPRESSION
see file proces.c

[18nov2003 mt]: 2 preprocessor variables are defined: 1) PARALLELTASK_ as
a process number, 0 is the master; 2) NPARALLELTASKS_ as a total
number of processes. These variables allow ParFORM to interact with
the operational system via #system and #pipe

[19nov2003 mt]: preprocessor variable is defined: CMODULE_ is a
current module number, starting from 1.

[26nov2003 mt]: bagfixes:
1) Broadcasting from MASTER to SLAVES of PreProcessor variables that
have changed is moved to the beginning of each block. The
corresponding procedure is PF_InitRedefinedPreVars in
parallel.c. Note, compilation is performed INDEPENDENTLY on
AC.mparallelflag, so No if(AC.mparallelflag==PARALLELFLAG)!
ATTENTION! In the previous version there was an error! Variables must
be broadcasted independently on mparallelflag!!!  The procedure is
used to broadcast number of terms in expression for preprocessor
if-expression 'termsin',
2) Number of terms in the expression is broadcasted from the master
to slaves for preprocessor if-expression 'termsin'. The problem here
is that only the master process can decide how many terms the
expression contains. So we have to broadcast from master to all others
this value. The corresponding procedure PF_BroadcastNumberOfTerms
is in the file 'parallel.c'.

[28nov2003 mt]: if the environment variable PF_SYNC is set, the
beginning of each module will be synchronized compulsory. This is done
via invocation PF_InitRedefinedPreVars() even if AC.NumberOfRedefsInModule==0.
This feature is useful only for debugging.

[02dec2003]: Bagfix: the structure of attachment of PreProcessor variables that
have changed in slave was wrong, see strings 02dec2003 in parallel.c.

[12dec2003 mt] Preprocessor directives "addseparator" and
"rmseparator" adds/removes separator characters used to separate
function arguments.  Example
#define QQ "a|g|a"
#addseparator %
*Comma must be quoted!:
#rmseparator ","
#rmseparator |
#call H(a,a%`QQ')

[30jan2004 mt]
Conception of on/off parallel is changed: instead of two flags, PARALLELFLAG 
and NOPARLLELFALG, I introduce several bit falgs:

#define NOPARALLEL_DOLLAR 1
#define NOPARALLEL_STORE 2
#define NOPARALLEL_RHS 4
#define NOPARALLEL_MOPT 8
#define NOPARALLEL_USER 16 
#define NOPARALLEL_TBLDOLLAR 32
#define PARALLEL_MOPT 64

#define PARALLELFLAG 0

So, now the resaon why execution can't be done in parallel will be reported 
to the user.

*************** Start further ChangeLog *****************
[19feb2004 mt]:
After introducing a differentiation in PARALLEL modes, sometimes extra 
warnings in sequential mode occur. To prevent such warnins I inserted
#ifdef PARALLEL in execute.c.

[19mar2004 mt]:
Large file support  is repaired (it was broken!), see comments in the file 
declare.h

[05apr2004 mt] - [07apr2004 mt] New directive:
cleartable thetable;
removes an existing table. If there is no such a table, the warning is generated:
"should be a table! -- Ignored".
And, [07apr2004 mt] -- memory leak is fixed in names.c:ResetVariables.

[28apr2004 mt]: 
signals SIGSEGV SIGFPE SIGILL SIGEMT SIGSYS SIGPIPE
SIGLOST SIGXCPU SIGXFSZ SIGTERM SIGINT SIGHUP SIGALRM SIGVTALRM
SIGPROF are captured in order to invoke Terminate(-1) on these signals.
To allow it at compile time, a macro
#define TRAPSIGNALS
should be defined in the file unix.h

[14apr2004 mt] - [27jul2004 mt] A new mechanism:
A new mechanism of interaction with external commands is implemented,
see an example below.

To allow the mechanism at compile time, a macro
#define WITHEXTERNALCHANNEL
should be defined in the file unix.h

INTRODUCTION

The preprocessor directives #system and #pipe execute the external
command synchronously i.e. they wait for the command to be completed.

The new directive #external starts the external command opening an
input-output channel for it, and can be used in order to extend FORM
in the spirit of a component model.

After the directive #external returns the control to the FORM program,
the external program initiated by the directive continues to run. Its
standard input and output references are stored by FORM. Both input
and output are not connected with any terminal device.

The directive #rmexternal terminates the external program. It
closes the programs' IO channels, sends to the program a KILL
signal and waits for the external program to be finished.                

The directive #toexternal is used to send some text to the running 
external program, the synthax is similar to the #write directive.


check/form.rb
# this file is to be included (required) in the actual tests

require 'open3'     # for popen3() to run FORM
require 'test/unit' # for the unit test classes
if RUBY_VERSION =~ /1\.8\..+/ # only for version 1.8.x
	require 'test/unit/ui/console/testrunner'
end

# names of the FORM executables
if ENV['bindir']
	FORM = ENV['bindir'] + "/form"
	TFORM = ENV['bindir'] + "/tform"
else
	FORM = "form"
	TFORM = "tform"
end

# maximal running time in seconds of FORM jobs before they get terminated
TIMEOUT = 10

# names of temporary files
TMPFILES_BASENAME = "form_rb_exam"
TESTPRG_NAME = TMPFILES_BASENAME+".frm"
LOGFILE_NAME = TMPFILES_BASENAME+".log"
STRACETMP_NAME = TMPFILES_BASENAME+"_strace"

def cleanup_pidfiles
	Dir.glob(STRACETMP_NAME+".*") { |fn| File.delete(fn) }
end
def cleanup_tempfiles
	File.delete(TESTPRG_NAME) if File.exist?(TESTPRG_NAME)
	File.delete(LOGFILE_NAME) if File.exist?(LOGFILE_NAME)
	cleanup_pidfiles
end

class FormTest < Test::Unit::TestCase
	def initialize(test_method_name)
		super(test_method_name)
		@input_file = ""
		@extra_parameter = ""
		@interactive = false
		cleanup_pidfiles
	end

	# Functions to specify test-run input and environment
	def input(str)
		File.open(TESTPRG_NAME, "w") { |f| f.write(str) }
		@input_file = TESTPRG_NAME
	end
	def input_file(str)
		@input_file = str
	end
	def extra_parameter(str)
		@extra_parameter = str
	end
	def interactive(true_or_false)
		@interactive = true_or_false
	end

	# Automatically called at end of each test run
	def teardown
		remove_files
	end
	def remove_files
		cleanup_tempfiles if passed?
	end

	# To run a (T)FORM job
	def execute(executable, timeout=TIMEOUT)
		cmdline = "strace -ff -o "+STRACETMP_NAME+" "+executable+" "+@extra_parameter+" "+@input_file
		@stdout = []; @stderr = []
		begin
			runner = Thread.current
			killer = Thread.new(timeout) { |timeout| sleep timeout; runner.raise }
			@inputstream, @outputstream, @errorstream = Open3.popen3(cmdline)
			out = Thread.new { while (( line = @outputstream.gets )) do @stdout << line end }
			err = Thread.new { while (( line = @errorstream.gets )) do @stderr << line end }
			out.join
			err.join
			killer.kill
		rescue
			@unfinished = true
		else
			@unfinished = false
		ensure
			@inputstream.close
			@outputstream.close
			@errorstream.close
			@stdout = @stdout.join
			@stderr = @stderr.join
			pids = []
			Dir.glob(STRACETMP_NAME+".*") { |fn| pids << File.extname(fn)[1..-1] }
			if pids.size == 0 then raise "Error: could not run executable #{executable}!" end
			pid = pids.sort[0]
			if @unfinished
				puts "Timeout: Terminating (T)FORM job!"
				system("kill -SIGKILL "+pid)
			end
		end
		evaluate pid
	end

	# Prepares data for test-result functions
	def evaluate(pid)
		File.open(STRACETMP_NAME+"."+pid, "r") { |f| @strace_out = f.readlines.join }
	end

	# Test-result functions 
	def return_value
		@strace_out =~ /exit_group\((\d+)\)/
		$1.to_i
	end
	def exact_result(exprname, index=-1) # verbatim result (keeps line breaks and whitespaces)
		matches = @stdout.scan(/^[ \t]+#{exprname}\s*=.+?;/m) 
		return matches[index] if not matches.empty?
		return ""
	end
	def result(exprname, index=-1) # result on one line with multiple whitespaces reduced to one
		exact_result(exprname, index).gsub(/\s+/, "")
	end

	# Test-result functions to be used in assertions
	def crash
		@strace_out =~ /Segmentation fault/
	end
	def no_crash; !crash; end
	def warning
		@stdout =~ /--> Warning:/
	end
	def no_warning; !warning; end
	def compile_error
		@stdout =~ /#{TESTPRG_NAME} Line \d+ -->/
	end
	def no_compile_error; compile_error; end
	def runtime_error
		@stdout =~ /Program terminating at #{TESTPRG_NAME} Line \d+ -->/
	end
	def no_runtime_error; !runtime_error; end
	def error
		compile_error || runtime_error || @stderr != ""
	end
	def no_error; !error; end
	def has_not_finished
		@unfinished
	end
	def has_finished
		!@unfinished
	end
	def problem
		has_not_finished || warning || error || crash
	end
	def no_problem; !problem; end

	# Utility functions for pattern matching
	def exact_pattern(str)
		san_str = Regexp.quote(str)
		Regexp.new(san_str)
	end
	def pattern(str)
		san_str = Regexp.quote(str.gsub(/\s+/, ""))
		Regexp.new(san_str)
	end
end

if __FILE__ == $0
	argv = ARGV
	options = argv.index("-c")
	$onlyoneclass = nil
	if options
		if options+1 >= argv.size
			puts "Error: Missing class name for option -c!"
			exit
		end
		$onlyoneclass = argv[options+1]
		argv.delete_at(options)
		argv.delete_at(options)
	end
	if argv == []
		if ENV['srcdir']
			path = ENV['srcdir'] + "/**/*.rb"
		else
			path = "**/*.rb"
		end
		files = Dir.glob(path)
	else
		files = argv
	end
	files.each do |iname|
		if iname != $0
			require iname
		end
	end
	if RUBY_VERSION =~ /1\.8\..+/
		class DerivedTests
			def self.suite
				suite = Test::Unit::TestSuite.new("(T)FORM tests")
				Class.constants.each do |c|
					cl = Class.class_eval(c)
					if cl.class == Class && cl.ancestors[1] == FormTest
						next if $onlyoneclass && $onlyoneclass != cl.name
						suite << cl.suite
					end
				end
				return suite
			end
		end
		Test::Unit::UI::Console::TestRunner.run(DerivedTests)
	elsif $onlyoneclass
		MiniTest::Unit::TestCase.reset
		MiniTest::Unit::TestCase.inherited(Kernel.const_get($onlyoneclass))
	end
end
